### DISCUSSION TITLE
Generate readable / compilable diffs

### DISCUSSION THREAD

**Discussion by Qottmann** [typst/typst#2211] [Q&A]
I am at a late stage of writing a manuscript in typst and I want to be able to create a diff that somehow highlights the differences between two versions of the manuscript. Is this possible, are there any tools compatible with typst?

**Comment by 8LWXpg** [comment]
writing typst in vscode then use git maybe?

**Reply by johannes-wolf** [reply]
I think what is meant is something like the output of `diff a.typ b.typ` applied to a Typst document, so that you see the differences in a compiled PDF.

**Reply by Qottmann** [reply]
Exactly, that would be ideal

**Comment by alerque** [comment]
This could be doable already with external tooling. I have a similar workflow I use with [SILE](https://sile-typesetter.org/). I use Markdown input that gets converted to SIL-markup wia `pandoc`. I wrote some tooling bundled in [CaSILE](https://github.com/sile-typesetter/casile) that generates a PDF diff between two Git branches of Markdown sources. It does this by first using `git diff` to compare the Markdown. Specifically I ask for the diff at the **word** level instead of line or character levels. This is the most useful result for the prose I usually work with. Usually this would generate a diff with terminal codes, but you can request how the diff markup be formatted. A little bit of coercing the output nets a Markdown document that uses [CriticMarkup](https://github.com/CriticMarkup/CriticMarkup-toolkit). Since my Markdown→SIL conversion supports CriticMarkup, what I get is a SIL input document with all the word changes tagged that can be styled with red/green/strikeout/whatever styling when typeset.

CriticMarkup is probably not relevant for Typst, but it should be pretty easy to workup an addon package with commands for formatting the components of a diff (removed, added, changed), and something similar to [`pandiff`](https://github.com/davidar/pandiff) (or even an extension of that) could be used to markup a diff using that package's commands.

**Reply by Qottmann** [reply]
Can you be more specific and include instructions / MWE to do this? I dont know either of these packages but potentially a blind copy-paste solution could do

**Reply by alerque** [reply]
This is notes on how I accomplished the same task using a different typesetting system that also didn't have built in support for the job. It is not something you can copy/paste, it is a workflow and pointers to tooling concepts that you would have to develop yourself.

**Comment by xxxserxxx** [comment]
> CriticMarkup is probably not relevant for Typst

It's a really good reference, though; it's well thought-through, and is a formalized specification for representing diffs in human-readable text. You're right in that CriticMarkup is focused on the markup, rather than providing a sort of formal specification that can be generalized, but it's a good starting point.  Converting the markup to `#add[]` `#remove[]` and `#substitute[]` (e.g.) depending on whether it's a block or a word-sized change would be a great start for basic diffs.

Doing diffs based on the git diff is less ideal than diffing the in-memory AST; I think this is how PanDiff works.

Edit: Actually, doing it with git is going to be impossible in any case, since any useful markup will have to be aware of Typst syntax.

**Reply by alerque** [reply]
> doing it with git is going to be impossible in any case

I get what you are saying here, and yes for best effect you'll want a diff engine that understands Typst syntax enough to break diff chunks at sensible boundaries. The caveat to add though is that Git itself can be configured to use external tooling to generate/render/manipulate diffs, so an ideal tool for this would function as a Git diff processor so CLI based diffing, merging, etc. could be made aware of the syntax as well as any rendered output versions.

**Reply by xxxserxxx** [reply]
How would those of us who don't use git manage it? Would this solution force us into a specific VCS? 

**Edit**: To be clear, I don't think this functionality necessarily belongs in the typst command itself. This sort of issue comes up in every markup language. Having it be external would be fine, but I **do** believe that it should be done at the AST level, and to support that, it'd be best to expose the internal AST of typst through a library, LSP, TreeSitter, or some structured document language like XML or JSON.  Just being able to dump and load the AST through the command line would be a big step to making a semantically-aware differ, and thereby, a diff-marked-up tool.

I will add that the Typst toolkit's interactive editor does suggest a more embedded approach, as edit tracking is a common functionality in WYSIWYG editors.

**Comment by DevelAngel** [comment]
I have the same problem: I want to generate a diff of two documents so that the reviewer does not need to look through the whole document. The reviewer is interested on text changes, not on layout changes.

To solve that problem, I checked pdf comparison tools like DiffPdf. But they does not fulfill all requirements: I need CLI support to be able to generate the diff in a CI/CD. Or the diff has a bad resolution.

Then, I looked into the generated Typst AST by using the typst-syntax crate. The AST contains variables, functions, module imports so that I think it is very hard to retrieve the real difference. I think, a reduced AST where all imports and functions are evaluated would be fine. I could not find a way to do this.

Next, I put some `eprintln!("{:#?}", ..)` in the typst compiler code to get an idea: I find out that the `Content` structure is the evaluated form of the AST.

Thus, I am currently writing a CLI to compare two typst documents which generates a pdf with highlighted texts by comparing two Content sequences. Another idea would be to generate a typst document instead of the pdf.

If I have something useful (I do this in my free time), I write it here...

By the way, my compliment to the Typst team and contributors: the documentation and code is very clean, well-structured and understandable. 👍

**Reply by mhdadk** [reply]
Thank you very much! This is exactly what I'm looking for. If you happen to come up with something, could you please post it in the Typst discord server [here](https://discord.com/channels/1054443721975922748/1204511855230324799/1204511855230324799)?

**Reply by JeppeGivskud-Cob** [reply]
Hey DevelAngel do you have any traction on your project? This would really take Typst to the next level :D

**Comment by falko** [comment]
I found the following projects when searching for this topic but have not (yet) tried them:
- [Typst](https://typst.app/) + [PanDiff](https://github.com/davidar/pandiff): [Gist](https://gist.github.com/Dragonink/2c50e4b9ed8ba071060f15b1977aa05c) [Discord](https://discord.com/channels/1054443721975922748/1339569197394694204/1339726123017310351) 
- https://github.com/kfijalkowski1/typst-diff-tool
- https://github.com/DevelAngel/typst-diff/tree/develop

**Reply by falko** [reply]
Issue https://github.com/typst/typst/issues/3049 also discusses this topic

**Comment by Atreyagaurav** [comment]
I know it's easy and people probably already have this made easily, but here is a snippet if someone wants to copy.

```typ
#let add(new) = text(blue, new)
#let rem(old) = strike(text(red, old))
#let diff(old, new) = {
    rem(old)
    add(new)
}
```

This is easy to use, (if you have a diff engine that takes custom patterns for how to format diff, it would be the best).

I do have a [program I use to generate diff](https://github.com/Atreyagaurav/difftext), but I made this to respond to LaTeX changes for an academic paper, so it takes latex input and generates typst. But it should work with typst inputs as well for the interactive mode. I can, in the future, expand this to take any input, with more user customization, and maybe also take a git-diff file for text based diffing if I have time or need of those features.

I understand text difference only is not enough for typst, because there can be dynamically generated contents. So for that, as others have said, a diff engine that understands typst and evaluates it is necessary. I would like to offer my help if someone is working on that. I think it is a valuable feature that'll make typst even more attractive to people.

I can think of two modes for this:
- `typst watch` like command (or a flag) where the edits since the first compile are highlighted. Very useful while working on a document. You can cancel it and compile it one last time when you are done.
- `typst diff` command that takes two typst files and generates the differences as an output (typst or PDF/SVG/PNG)

**Reply by istudyatuni** [reply]
You may want to share it in the `#showcase` channel on discord and/or on forum

**Reply by Atreyagaurav** [reply]
Maybe once I add more features, right now it's really only useful if someone takes code from it and makes something for themselves.
